[LoongArch] Emit function call code sequence as `PCADDU18I+JIRL` in medium code model

wangleiat opened the pull request:
Adjustments made in accordance with psABI v2.30:
https://github.com/loongson/la-abi-specs/releases/tag/v2.30

1. Moved the expansion of relevant pseudo-instructions from `LoongArchPreRAExpandPseudo` pass to `LoongArchExpandPseudo` pass, to ensure that the code sequences of `PseudoLA*_LARGE` instructions and Medium code model's function call are not scheduled.

2. AsmParser now supports parsing for call36 and tail36 macro instructions.

wangleiat committed:
[LoongArch] Emit function call code sequence as `PCADDU18I+JIRL` in medium code model

According to the description of the psABI v2.20:
https://github.com/loongson/la-abi-specs/releases/tag/v2.20, adjustments
are made to the function call instructions under the medium code model.

At the same time, AsmParser has already supported parsing the call36 and
tail36 macro instructions.

wangleiat commented:
@xen0n @xry111

xen0n reviewed (commented):
Sorry for the late review and happy new year!

It seems the commit diff is a bit large though -- there's a whole refactoring (separating large code model logic into several different functions, plus code movement) being lumped together with the `call36` implementation. It could be easier to split into two PRs so each change stays focused. I've lightly checked the `call36` test case changes and they look good though.

wangleiat committed:
[LoongArch] Reimplement the expansion of PseudoLA*_LARGE instructions

According to the description of the psABI v2.30:
https://github.com/loongson/la-abi-specs/releases/tag/v2.30, moved the
expansion of relevant pseudo-instructions from
`LoongArchPreRAExpandPseudo` pass to `LoongArchExpandPseudo` pass, to
ensure that the code sequences of `PseudoLA*_LARGE` instructions and
Medium code model's function call are not scheduled.

wangleiat committed:
[LoongArch] Pre-commit test for xxx. NFC

wangleiat commented:
> Sorry for the late review and happy new year!
> 
> It seems the commit diff is a bit large though -- there's a whole refactoring (separating large code model logic into several different functions, plus code movement) being lumped together with the `call36` implementation. It could be easier to split into two PRs so each change stays focused. I've lightly checked the `call36` test case changes and they look good though.

Thank you very much for your suggestion. I have already implemented the split as you advised.

SixWeining reviewed (approved):
LGTM. These 3 commits should have to be merged one by one locally. And the 2nd one's title ("xxx") should be updated.

xen0n commented on llvm/lib/Target/LoongArch/LoongArchInstrInfo.td:
```diff
@@ -1479,28 +1548,30 @@ let hasSideEffects = 0, mayLoad = 1, mayStore = 0, isCodeGenOnly = 0,
+                             "la.tls.ld", "$dst, $src">;
+def PseudoLA_TLS_GD : Pseudo<(outs GPR:$dst), (ins bare_symbol:$src), [],
+                             "la.tls.gd", "$dst, $src">;
+let Defs = [R20], Size = 20 in {
```
Would it be possible to let the register allocator choose the temp? Right now as `$t8` is temporary it's fine though.

wangleiat replied:
> Would it be possible to let the register allocator choose the temp? Right now as `$t8` is temporary it's fine though.

It should be possible, and it may require defining some nodes to achieve it. There will be more changes in the `TargetLowering` part, which can be left for subsequent tasks.

xen0n reviewed (commented):
The test case name `no-schedue` could be something more descriptive like "psabi-restricted-scheduling".

Overall it's mostly good and patch no. 1 is LGTM, thanks!

SixWeining commented on llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchFixupKinds.h:
```diff
@@ -108,7 +108,10 @@ enum Fixups {
+  fixup_loongarch_relax = FirstLiteralRelocationKind + ELF::R_LARCH_RELAX,
+  // 36-bit fixup corresponding to %call36(foo) for a pair instructions:
+  // pcaddu18i+jirl.
+  fixup_loongarch_call36 = FirstLiteralRelocationKind + ELF::R_LARCH_CALL36
```
Nit: Better to add a `,` at the end of this line to avoid touching it in future when we add new fixups.

wangleiat replied:
> Nit: Better to add a `,` at the end of this line to avoid touching it in future when we add new fixups.

Thank you very much. I will make the corrections during the merge.

wangleiat commented:
In order to preserve each commit message, the submission was done using the command line.
